src/compiler/checker.ts

```diff
@@ -4092,7 +4106,15 @@ namespace ts {
                         const resolvedSymbol = resolveName(param, paramSymbol.name, SymbolFlags.Value, undefined, undefined);
                         paramSymbol = resolvedSymbol;
                     }
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {
+                        thisType = param.type && getTypeOfSymbol(paramSymbol);

```

Comment:
So, if the programmer omits a type annotation on a `this` parameter, you will think there was *no* `this` parameter and get your parameter count adjustments wrong below. You need to track the presence of a this parameter separately from its type. When the type annotation is missing you should use `unknownType`. Also, you can just use `getTypeFromTypeNode(param.type)` instead of `getTypeOfSymbol`.